src/Data/Maybe.ard

```diff
@@ -0,0 +1,52 @@
+\import Logic
+\import Paths
+\import Set
+
+-- | Maybe A
+-- Maybe represents an optional value. It might also be used to represent failure.
+\data Maybe (A : \Type) | Nothing | Just A
+  \where {
+    \func bind {A B : \Type} (a : Maybe A)(f : A -> Maybe B) : Maybe B \elim a
+      | Nothing => Nothing
+      | Just a => f a
+
+    \lemma JustA/=Nothing  {A : \Type} (a : A) (p :  Just a = Nothing) : Empty =>
+      IsJust.absurd (transport IsJust p ItIsJust)
+  }
+
+
+-- | IsJust is a proof that the value is actually a Just
+\data IsJust {A : \Type} (m : Maybe A) : \Prop \with
+  | Just _ => ItIsJust
+  \where {
+    \func absurd {A B : \Type} (s : IsJust (Nothing{A})) : B
+
+    \lemma notNothing {A : \Type} (m : Maybe A) (ev : IsJust m) : Not (m = Nothing) \elim m, ev
+      | Just a, ItIsJust => \lam x => Maybe.JustA/=Nothing a x
+  }
+
+-- | IsNothing is a proof that maybe is Nothing
+\data IsNothing {A : \Type} (Maybe A) : \Prop \with
```

Comment:
\func IsNothing {A : \Type} (m : Maybe A) => m = Nothing` and I consider it unnecessary